blockprod/src/detail/mod.rs

```diff
@@ -254,12 +260,47 @@ impl BlockProduction {
             self.job_manager.make_job_stopper_function();
         let _job_stopper_destructor = OnceDestructor::new(move || job_stopper_function(job_key));
 
+        // Unlike Proof of Work, which can vary any header field when
+        // searching for a valid block, Proof of Stake can only vary
+        // the header timestamp. Its search space starts at the
+        // previous block's timestamp + 1 second, and ends at the
+        // current timestamp + some distance in time defined by the
+        // blockchain.
+        //
+        // This variable keeps track of the last timestamp that was
+        // attempted, and during Proof of Stake, will prevent
+        // searching over the same search space.
+        let last_epoch_used = {
+            let tip_timestamp = tip_at_start.block_timestamp();
+
+            let tip_plus_one = tip_timestamp
+                .add_int_seconds(1)
+                .ok_or(ConsensusCreationError::TimestampOverflow(tip_timestamp, 1))?;
+
+            Arc::new(AtomicU64::new(tip_plus_one.as_int_seconds()))
+        };
+
+        let current_timestamp =
```

Comment:
As a nitpick, since this is called only once, you can scope it inside whatever calculates `max_block_timestamp`

Reply:
Yep, I like doing that. Nice catch.